src/bot.ts

```
@@ -0,0 +1,36 @@
+import { config } from "./config.ts";
+
+import { Bot } from "grammy";
+
+const bot = new Bot(config.BOT_TOKEN);
+
+/**
+ * @param {number} gpa - TODO
+ * @returns {number} TODO
+ */
+function calculateScholarship(gpa: number): number {
+  if (!(2 <= gpa && gpa <= 5)) {
+    throw new RangeError(/* TODO */);
+  }
+
+  const M_MIN = 3000;
+  const M_MAX = 20_000;
+
+  let s = M_MIN + (M_MAX - M_MIN) * ((gpa - 2) / 3) ** 2.5;
+  s = Math.floor(s / 100) * 100;
+
+  return s;
+}
+
+bot.on("message:text", async (ctx) => {
+  const gpa = parseFloat(ctx.message.text);
+
+  try {
+    const scholarship = calculateScholarship(gpa);
+    await ctx.reply(scholarship.toString());
+  } catch (_) {
+    ctx.reply("Invalid input!");
+  }
```

Remove try/catch and validate user input explicitly

---

Then the program will crash in case of wrong input.

---

You should check that `gpa` and reply to user that it is wrong, if it's.
If error will be thrown after validation, then there is actually an error in program and it should throw.

---

But if I don't validate the value in the function, won't it be unsafe? If the function receives incorrect data, it should either throw an error or return null/undefined/whatever. I cannot just assume that the `gpa` argument is always in the valid range. I could use another type for this argument, but it's pretty tough in TS to specify the range for the value.

---

you can use some specific exception instead of wildcart.

> But if I don't validate the value in the function, won't it be unsafe? If the function receives incorrect data, it should either throw an error or return null/undefined/whatever. I cannot just assume that the `gpa` argument is always in the valid range. I could use another type for this argument, but it's pretty tough in TS to specify the range for the value.

you can use some specific exception instead of wildcart. In Python, f.e. ValueError used for any validation-related stuff.

---

If I understand you correctly, that's exactly what I'm doing. The function throws RangeError if `gpa` is not in the range [2, 5]. Then each time calling this function I catch this error.

---

So, maybe specifying Exception type in catch statement will be appropriate? 

`try {
    // Code
} catch (e: any) {
    if (e instanceof YourException){
        // code that runs for this exception.
    } else if (e instanceof YourException2){
        // code that runs for this exception.
    } else {
       // Throw Runtime Error or something like that, I really don't know TS...
    }
}`

---

Also you can check how Artem has solved this: 
https://github.com/one-zero-eight/InNoHassle-Website/blob/main/components/ScholarshipCalculator.tsx

---

Actually, this seems to be the solution!